fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/api/transfer/v1/context/ContainerItemContext.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.transfer.v1.context;
+
+import java.util.List;
+
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.screen.ScreenHandler;
+import net.minecraft.util.Hand;
+import net.minecraft.world.World;
+
+import net.fabricmc.fabric.api.lookup.v1.item.ItemApiLookup;
+import net.fabricmc.fabric.api.transfer.v1.item.InventoryStorage;
+import net.fabricmc.fabric.api.transfer.v1.item.ItemVariant;
+import net.fabricmc.fabric.api.transfer.v1.storage.Storage;
+import net.fabricmc.fabric.api.transfer.v1.storage.StoragePreconditions;
+import net.fabricmc.fabric.api.transfer.v1.storage.base.SingleSlotStorage;
+import net.fabricmc.fabric.api.transfer.v1.transaction.Transaction;
+import net.fabricmc.fabric.api.transfer.v1.transaction.TransactionContext;
+import net.fabricmc.fabric.impl.transfer.context.PlayerContainerItemContext;
+import net.fabricmc.fabric.impl.transfer.context.SingleSlotContainerItemContext;
+
+/**
+ * A context that determines how an {@link ItemVariant} interacts with an inventory, or at least the part of it that is visible to the context.
+ * For example, it allows a water bucket to replace itself by an empty bucket when emptied.
+ * A {@code ContainerItemContext} generally represents a slot in an inventory, and additionally allows access to the rest of the inventory,
+ * which may or may not contain the preferred slot itself.
+ *
+ * <p>When an {@linkplain ItemApiLookup item API} requires a {@code ContainerItemContext} as context,
+ * it will generally be suitable to obtain a context instance with {@link #ofPlayerHand} or {@link #ofPlayerCursor},
+ * and then use {@link #find} to query an API instance.
+ *
+ * <p>Implementors of item APIs will generally want to:
+ * <ul>
+ *     <li>Query how much of the variant is in the inventory through {@link #getAmount}.</li>
+ *     <li>Transform some of the bound variant into another variant through {@link #transform}.</li>
+ * </ul>
+ */
+// TODO: update javadoc
+// TODO: wrapper around a SingleSlotStorage<ItemVariant>
+// TODO: wrapper around a StackReference maybe?
+public interface ContainerItemContext {
```

Comment:
Why doesn't this subclass Storage?

Reply:
I did another pass on this, have a look at the new design!